samples/rust/rust_seq_file.rs

```diff
@@ -0,0 +1,136 @@
+// SPDX-License-Identifier: GPL-2.0
+
+//! Example of using a [`seq_file`] in Rust.
+//!
+//! C header: [`include/linux/seq_file.h`](../../../include/linux/seq_file.h)
+//! C header: [`include/linux/proc_fs.h`](../../../include/linux/proc_fs.h)
+//!
+//! Reference: <https://www.kernel.org/doc/html/latest/filesystems/seq_file.html>
+
+#![no_std]
+#![feature(allocator_api, global_asm, try_reserve)]
+
+use alloc::{boxed::Box, sync::Arc};
+use core::{
+    convert::TryInto,
+    fmt::Write,
+    iter::{repeat, Peekable, Repeat, Take},
+    pin::Pin,
+};
+use kernel::{
+    cstr,
+    file_operations::{File, FileOpener, FileOperations},
+    io_buffer::IoBufferWriter,
+    miscdev, mutex_init,
+    prelude::*,
+    seq_file,
+    sync::Mutex,
+};
+
+module! {
+    type: RustSeqFileDev,
+    name: b"rust_seq_file",
+    author: b"Adam Bratschi-Kaye",
+    description: b"Rust sample using a seq_file",
+    license: b"GPL v2",
+    params: {
+    },
+}
+
+#[derive(Clone)]
+struct SharedState(Arc<Mutex<u32>>);
+
+impl SharedState {
+    fn try_new() -> KernelResult<Self> {
+        let state = Arc::try_new(
+            // SAFETY: `mutex_init!` is called below.
+            unsafe { Mutex::new(0) },
+        )?;
+        // SAFETY: Mutex is pinned behind `Arc`.
+        let pin_state = unsafe { Pin::new_unchecked(state.as_ref()) };
+        mutex_init!(pin_state, "SharedState::0");
+        Ok(SharedState(state))
+    }
+}
+
+impl seq_file::SeqOperations for SharedState {
+    type Item = String;
+    type Iterator = Take<Repeat<String>>;
+
+    fn start(arg: &SharedState) -> Option<Box<Peekable<Self::Iterator>>> {
+        let count = arg.0.lock();
+        let mut message = String::new();
+
+        let template = "rust_seq_file: device opened this many times: ";
+        message.try_reserve_exact(template.len() + 4).ok()?;
+        // NOPANIC: We reserved space for `template` above.
+        message.push_str(template);
+        if *count < 1000 {
+            // NOPANIC: There are 4 characters remaining in the string which
+            // leaves space for a 3 digit number and the newline.
+            write!(&mut message, "{}\n", *count).ok()?;
+        }
```

Comment:
Something more explicit for the 3, 4 and 1000 constants:

```rust
const MAX_DIGITS: u32 = 3;
const MAX_LENGTH: u32 = MAX_DIGITS + 1;
const MAX_COUNT: u32 = 10u32.pow(MAX_DIGITS);
```

Reply:
done.

Reply:
done

Reply:
I would put them local to the function, given they are not used elsewhere.

Reply:
:+1: